tests/st/policy/test_felix_gateway.py

```diff
@@ -0,0 +1,510 @@
+        # docker-in-docker subnet
+        log_and_run("docker network create --driver bridge --subnet 172.19.0.0/16 cali-st-ext")
+
+        # And an nginx server on the external network only.
+        log_and_run("docker run"
+                    " --network=cali-st-ext"
+                    " -d"
+                    " --name=cali-st-ext-nginx"
+                    " nginx")
+
+        for host in cls.hosts:
+            host.start_calico_node()
+
+        # Get the internal IP of the gateway.  We do this before we add the second
+        # network since it means we don't have to figure out which IP is which.
+        int_ip = cls.get_container_ip("cali-st-gw")
+        cls.gateway_int_ip = int_ip
+        _log.info("Gateway internal IP: %s", cls.gateway_int_ip)
+
+        # Add the gateway to the external network.
+        log_and_run("docker network connect cali-st-ext cali-st-gw")
+        cls.gateway.execute("ip addr")
+
+        # Get the IP of the external server.
+        ext_ip = cls.get_container_ip("cali-st-ext-nginx")
+        cls.ext_server_ip = ext_ip
+        _log.info("External workload IP: %s", cls.ext_server_ip)
+
+        # Configure the internal host to use the gateway for the external IP.
+        cls.host.execute("ip route add %s via %s" %
+                         (cls.ext_server_ip, cls.gateway_int_ip))
+
+        # Configure the gateway to forward and NAT.
+        cls.gateway.execute("sysctl -w net.ipv4.ip_forward=1")
+        cls.gateway.execute("iptables -t nat -A POSTROUTING --destination %s -j MASQUERADE" %
+                            cls.ext_server_ip)
+
+    def setUp(self):
+        # Override the per-test setUp to avoid wiping etcd; instead only clean up the data we
+        # added.
+        self.remove_pol_and_endpoints()
+
+    def tearDown(self):
+        self.remove_pol_and_endpoints()
+        super(TestFelixOnGateway, self).tearDown()
+
+    @classmethod
+    def tearDownClass(cls):
+        # Tidy up
+        for host in cls.hosts:
+            host.remove_workloads()
+        for host in cls.hosts:
+            host.cleanup()
+            del host
+
+        log_and_run("docker rm -f cali-st-ext-nginx || true")
+
+    def test_ingress_policy_can_block_through_traffic(self):
+        self.add_policy({
+            'apiVersion': 'v1',
+            'kind': 'policy',
+            'metadata': {'name': 'port80-int'},
+            'spec': {
+                'order': 10,
+                'ingress': [
+                    {
+                        'protocol': 'tcp',
+                        'destination': {'ports': [80]},
+                        'action': 'deny'
+                    },
+                ],
+                'egress': [
+                    {'action': 'deny'},
+                ],
+                'selector': 'role == "gateway-int"'
+            }
+        })
+        self.add_gateway_internal_iface()
+        retry_until_success(self.assert_host_can_not_curl_ext, 3)
```

Comment:
If you use the workload, you might also be able to use the existing (no) connectivity check functions to test TCP/UDP/Ping all at once.